File: demos/typescript/geometry-samples/src/samples/bvh.ts
// Bounding volume hierarchy (BVH) over mesh triangles.
//
// Recursively partition triangles by sorting their centroids along the longest
// axis of the current bounds and splitting at the median. The result is a
// balanced tree of axis-aligned boxes that takes a raycast or an intersection
// query from O(n) to O(log n).
//
// The box algebra is the stdlib's: `Bounds3D.Empty` as the identity for
// `UnionOfBounds`, `Bounds3D.Extent` for the split axis, `Triangle3D.Centroid`
// for the sort key. The stdlib also ships the QUERY half of a BVH
// (`Bvh3D.Raycast`, `CandidatesInBounds`), but nothing that builds one — see
// plato-442.
//
// `Triangle3D.Bounds()` is deliberately not used: it resolves to the Bounds2D
// overload of BoundsOfPoints and silently returns a 2D box (plato-441).

import type { Drawable, Sample } from '../core/types.js';
import { Bounds3D, Line3D, Point3D, Triangle3D, TriangleMesh3D, Vector3D } from '../plato/plato.g.js';
import { boxEdges, toArray, translateMesh } from '../core/meshBuilder.js';
import { buildIcosphere } from './icosphere.js';

export interface BvhNode {
    bounds: Bounds3D;
    depth: number;
    left?: BvhNode;
    right?: BvhNode;
    /** Triangle indices (leaves only). */
    triangles?: number[];
}

export const triangleCentroid = (triangles: Triangle3D[], t: number): Point3D =>
    triangles[t].Centroid();

/** The axis-aligned box of one triangle. */
const triangleBounds = (t: Triangle3D): Bounds3D =>
    [t.A, t.B, t.C].reduce(
        (b, p) => new Bounds3D(
            new Point3D(b.Min.X.Min(p.X), b.Min.Y.Min(p.Y), b.Min.Z.Min(p.Z)),
            new Point3D(b.Max.X.Max(p.X), b.Max.Y.Max(p.Y), b.Max.Z.Max(p.Z))),
        Bounds3D.Empty());

/** The union of a set of triangle boxes; Bounds3D.Empty is the identity. */
const boundsOf = (triangles: Triangle3D[], indices: number[]): Bounds3D =>
    indices.reduce((acc, t) => acc.UnionOfBounds(triangleBounds(triangles[t])), Bounds3D.Empty());

export function buildBvh(triangles: Triangle3D[], indices: number[], depth = 0, leafSize = 8): BvhNode {
    const bounds = boundsOf(triangles, indices);
    if (indices.length <= leafSize || depth >= 16)
        return { bounds, depth, triangles: indices };

    // Median split along the longest axis of the bounds.
    const extent = bounds.Extent();
    const axis: 'X' | 'Y' | 'Z' =
        extent.X >= extent.Y && extent.X >= extent.Z ? 'X' : extent.Y >= extent.Z ? 'Y' : 'Z';
    const sorted = indices.slice().sort((a, b) =>
        triangleCentroid(triangles, a)[axis] - triangleCentroid(triangles, b)[axis]);
    const half = sorted.length >> 1;

    return {
        bounds, depth,
        left: buildBvh(triangles, sorted.slice(0, half), depth + 1, leafSize),
        right: buildBvh(triangles, sorted.slice(half), depth + 1, leafSize),
    };
}

/** The boxes of every node at a given depth (leaves shallower than it included). */
export function boxesAtDepth(node: BvhNode, depth: number): Line3D[] {
    if (node.depth === depth || (node.triangles && node.depth < depth))
        return boxEdges(node.bounds.Min, node.bounds.Max);
    return [
        ...(node.left ? boxesAtDepth(node.left, depth) : []),
        ...(node.right ? boxesAtDepth(node.right, depth) : []),
    ];
}

/** Every triangle of a mesh, gathered by the stdlib. */
export const meshTriangles = (mesh: TriangleMesh3D): Triangle3D[] => toArray(mesh.Primitives());

export const bvhSample: Sample = {
    id: 'bvh',
    title: 'BVH (AABB Tree)',
    description: 'Median-split Bounds3D hierarchy over TriangleMesh3D.Primitives, at three depths.',
    build(): Drawable[] {
        const mesh = buildIcosphere(3);
        const triangles = meshTriangles(mesh);
        const root = buildBvh(triangles, triangles.map((_, i) => i));

        const colors = [0xff6b6b, 0xffd166, 0x6fbf73];
        const panels = [2, 4, 6].map((depth, i): Drawable[] => {
            const shift = new Vector3D((i - 1) * 2.6, 0, 0);
            const move = (p: Point3D) => p.Add(shift);
            return [
                {
                    kind: 'mesh',
                    mesh: translateMesh(mesh, shift),
                    color: 0x4da3ff,
                    opacity: 0.25,
                },
                {
                    kind: 'lines',
                    segments: boxesAtDepth(root, depth)
                        .map(s => new Line3D(move(s.A), move(s.B))),
                    color: colors[i],
                    opacity: 0.85,
                },
            ];
        });

        return panels.flat();
    },
};

File: demos/rust/geometry-samples/src/samples/bvh.rs
// Bounding volume hierarchy (BVH) over mesh triangles.
//
// Recursively partition triangles by sorting their centroids along the
// longest axis of the current bounds and splitting at the median. The result
// is a balanced tree of axis-aligned boxes used to accelerate raycasts and
// intersection queries from O(n) to O(log n).

use crate::core::mesh_builder::{append_box_edges, vertex_at};
use crate::core::types::{Drawable, MeshData};
use crate::plato::*;

use super::icosphere::build_icosphere;

pub struct BvhNode {
    pub min: Vector3D,
    pub max: Vector3D,
    pub depth: usize,
    pub left: Option<Box<BvhNode>>,
    pub right: Option<Box<BvhNode>>,
    /// Triangle indices (leaves only).
    pub triangles: Option<Vec<usize>>,
}

pub fn triangle_centroid(mesh: &MeshData, t: usize) -> Vector3D {
    let a = vertex_at(&mesh.positions, mesh.indices[t * 3] as usize);
    let b = vertex_at(&mesh.positions, mesh.indices[t * 3 + 1] as usize);
    let c = vertex_at(&mesh.positions, mesh.indices[t * 3 + 2] as usize);
    a.Add(b).Add(c).Scale(1.0 / 3.0)
}

fn triangle_bounds(mesh: &MeshData, t: usize) -> (Vector3D, Vector3D) {
    let mut min = Vector3D::new(f64::INFINITY, f64::INFINITY, f64::INFINITY);
    let mut max = Vector3D::new(f64::NEG_INFINITY, f64::NEG_INFINITY, f64::NEG_INFINITY);
    for k in 0..3 {
        let v = vertex_at(&mesh.positions, mesh.indices[t * 3 + k] as usize);
        min = min.Min(v);
        max = max.Max(v);
    }
    (min, max)
}

pub fn build_bvh(mesh: &MeshData, triangles: Vec<usize>, depth: usize, leaf_size: usize) -> BvhNode {
    let mut min = Vector3D::new(f64::INFINITY, f64::INFINITY, f64::INFINITY);
    let mut max = Vector3D::new(f64::NEG_INFINITY, f64::NEG_INFINITY, f64::NEG_INFINITY);
    for &t in &triangles {
        let (bmin, bmax) = triangle_bounds(mesh, t);
        min = min.Min(bmin);
        max = max.Max(bmax);
    }

    if triangles.len() <= leaf_size || depth >= 16 {
        return BvhNode { min, max, depth, left: None, right: None, triangles: Some(triangles) };
    }

    // Median split along the longest axis of the bounds.
    let extent = max.Subtract(min);
    let axis = if extent.X >= extent.Y && extent.X >= extent.Z {
        0
    } else if extent.Y >= extent.Z {
        1
    } else {
        2
    };
    let component = |v: Vector3D| match axis {
        0 => v.X,
        1 => v.Y,
        _ => v.Z,
    };
    let mut sorted = triangles;
    sorted.sort_by(|&a, &b| {
        component(triangle_centroid(mesh, a))
            .partial_cmp(&component(triangle_centroid(mesh, b)))
            .unwrap()
    });
    let half = sorted.len() / 2;
    let right_tris = sorted.split_off(half);

    BvhNode {
        min,
        max,
        depth,
        left: Some(Box::new(build_bvh(mesh, sorted, depth + 1, leaf_size))),
        right: Some(Box::new(build_bvh(mesh, right_tris, depth + 1, leaf_size))),
        triangles: None,
    }
}

pub fn collect_boxes_at_depth(node: &BvhNode, depth: usize, out: &mut Vec<f64>) {
    if node.depth == depth || (node.triangles.is_some() && node.depth < depth) {
        append_box_edges(node.min, node.max, out);
        return;
    }
    if let Some(left) = &node.left {
        collect_boxes_at_depth(left, depth, out);
    }
    if let Some(right) = &node.right {
        collect_boxes_at_depth(right, depth, out);
    }
}

pub fn build() -> Vec<Drawable> {
    let mesh = build_icosphere(3);
    let all_tris: Vec<usize> = (0..mesh.indices.len() / 3).collect();
    let root = build_bvh(&mesh, all_tris, 0, 8);

    let mut drawables: Vec<Drawable> = Vec::new();

    for k in [-1.0f64, 0.0, 1.0] {
        let mut m = build_icosphere(3);
        for j in (0..m.positions.len()).step_by(3) {
            m.positions[j] += k * 2.6;
        }
        drawables.push(m.into_drawable());
    }

    for (i, depth) in [2usize, 4, 6].iter().enumerate() {
        let mut positions: Vec<f64> = Vec::new();
        collect_boxes_at_depth(&root, *depth, &mut positions);
        for j in (0..positions.len()).step_by(3) {
            positions[j] += (i as f64 - 1.0) * 2.6; // side by side
        }
        drawables.push(Drawable::Lines { positions });
    }

    drawables
}
